src/realm/realm.h

```diff
@@ -0,0 +1,1612 @@
+/*
+    FIXME: License, since this header may be distributed independently from
+    other headers.
+*/
+
+#ifndef REALM_H
+#define REALM_H
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdbool.h>
+
+#if defined(_WIN32) || defined(__CYGWIN__)
+
+#if defined(Realm_EXPORTS)
+// Exporting Win32 symbols
+#define RLM_EXPORT __declspec(dllexport)
+#else
+// Importing Win32 symbols. Note: Clients linking statically should define
+// RLM_NO_DLLIMPORT.
+#if !defined(RLM_NO_DLLIMPORT)
+#define RLM_EXPORT __declspec(dllimport)
+#else
+#define RLM_EXPORT
+#endif // RLM_NO_DLLIMPORT
+#endif // Realm_EXPORTS
+
+#else
+// Not Win32
+#define RLM_EXPORT __attribute__((visibility("default")))
+#endif
+
+#ifdef __cplusplus
```

Comment:
The drawback of that is that the compiler will not check whether a function definition in C++ code matches the C declaration unless the definition is also marked `extern "C"`, and it would be unfortunate to have huge `extern "C"` blocks in the cpp files as well.

Reply:
-Wmissing-prototypes can catch that.

Reply:
You don't need `extern "C"` in the function definition.

Reply:
You don't, but as I said it has an important benefit.

`-Wmissing-prototypes` detects calling a function without any declaration, which is legal in C (but not C++).

`-Wmissing-declarations` could be used to the same effect, but I still prefer the very short `RLM_API` prefix to huge `extern "C"` blocks, which are annoying for the same reasons that huge `namespace {}` blocks are annoying.

Reply:
(We can do both, though.)

Reply:
I'm resolving this in the service of choosing a color for the bike shed. :)